Allow minimized attribute completions anywhere in the attribute

davidwengier opened the pull request:
Fixes https://github.com/dotnet/razor-tooling/issues/6578

davidwengier commented on src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/Completion/TagHelperCompletionProviderTest.cs:
```diff
@@ -402,7 +402,34 @@ public void GetCompletionAt_AtAttributeEdge_IndexerAttribute_ReturnsCompletions(
+                },
+                completion =>
+                {
+                    Assert.Equal("int-val", completion.InsertText);
```
My knowledge of completion here is lacking. Should this be here? It's not shown because the IDE filters it out, but maybe we're supposed to do the filtering??

davidwengier commented:
I tested manually with the Title attribute of Survey prompt, but I'll double check that case too when I'm next at a computer. Thanks for the suggestion.

DoctorKrolic commented:
> I tested manually with the Title attribute of Survey prompt

Did you test case, when there is a `Title` value as I showed in the issue, e.g. `<SurveyPrompt Tit$$le="Test />"`. I have a suspicion is that this can break because only minimalized attributes are considered. Would like to see such test in the codebase.

davidwengier commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/TagHelperCompletionProvider.cs:
```diff
@@ -123,11 +123,13 @@ static bool IsPossiblePartiallyWrittenAttribute(SyntaxNode? attributeSyntax, int
                     if (attributeSyntax is MarkupMinimizedTagHelperAttributeSyntax or MarkupMinimizedAttributeBlockSyntax)
                     {
-                        return attributeSyntax.Span.End == absoluteIndex;
+                        return attributeSyntax.Span.Start < absoluteIndex && attributeSyntax.Span.End >= absoluteIndex;
```
That won't compile I'm afraid, patterns only work with compile time constants.

DoctorKrolic replied:
Uh, probably yes, sorry for distracting with this

davidwengier replied:
No problem at all! I'm just sitting on the couch watching a crappy Netflix show, distractions are welcome :)

davidwengier commented:
> Would like to see such test in the codebase.

Thank you, I missed this scenario, and in fact there is a test that covers it but of course it is expecting the wrong thing. Works like this with the fix:

![TitleCtrlSpace](https://user-images.githubusercontent.com/754264/179635275-c40dee06-eff5-4f0b-9683-49b3a340815c.gif)

davidwengier commented:
Expanded this to more attribute types, and updated the tests. The tests that I've changed don't seem problematic to me, but I might be missing something. It appears the previous logic was added in https://github.com/dotnet/razor-tooling/pull/3839 but that PR and linked issue only talks about tags, not attributes, and I haven't changed them, so maybe the restriction on attributes was less intentional? @NTaylorMullen?

DoctorKrolic commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/TagHelperCompletionProvider.cs:
```diff
@@ -119,20 +119,16 @@ public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorCompl
-                    return false;
+                    // We allow the absoluteIndex to be anywhere in the attribute, and for non minimized attributes,
+                    // so that `<SurveyPrompt Title=""` doesn't return only the html completions, because that has the effect of overwriting the casing of the attribute.
+                    return attributeSyntax is MarkupMinimizedTagHelperAttributeSyntax or MarkupMinimizedAttributeBlockSyntax or MarkupTagHelperAttributeSyntax &&
```
What about non-tag helper full attributes? See my comment above: if a tag helper has restriction, so it becomes a tag helper only with certain attributes, will it work as expected in this case?

davidwengier replied:
Your "class" case? I got no completions for that, as expected

NTaylorMullen commented:
> It appears the previous logic was added in https://github.com/dotnet/razor-tooling/pull/3839 but that PR and linked issue only talks about tags, not attributes, and I haven't changed them, so maybe the restriction on attributes was less intentional? @NTaylorMullen?

So the intention was for all completions (including attributes) but this was under the impression that HTML similarly would not return completions in "middle of word" for attributes. Given that seems to not be true this falls in line with what we'd want.

ryanbrandenburg commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/TagHelperCompletionProvider.cs:
```diff
@@ -119,18 +119,16 @@ public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorCompl
                     // When we are in the middle of writing an attribute it is treated as a minimilized one, e.g.:
                     // <form asp$$ - 'asp' is parsed as MarkupMinimizedTagHelperAttributeSyntax (tag helper)
-                    // <SurveyPromt Tit$$ - 'Tit' is parsed as MarkupMinimizedTagHelperAttributeSyntax as well (razor component)
+                    // <SurveyPrompt Tit$$ - 'Tit' is parsed as MarkupMinimizedTagHelperAttributeSyntax as well (razor component)
```
❤️

ryanbrandenburg commented on src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/TagHelperCompletionProvider.cs:
```diff
@@ -119,18 +119,16 @@ public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorCompl
                 return attributeCompletions;
 
-                static bool IsPossiblePartiallyWrittenAttribute(SyntaxNode? attributeSyntax, int absoluteIndex)
+                static bool InOrAtEndOfAttribute(SyntaxNode attributeSyntax, int absoluteIndex)
```
Much better name too.

davidwengier commented:
Thanks for the reviews and comments. Gonna merge it, but happy to follow up if more issues are found. I'm sure you're all better at finding them than me :)
